### ISSUE TITLE
RFC: Improved YAML encoding

### ISSUE THREAD

**Issue by ntijanic** [common-workflow-language/common-workflow-language#173]
[Previous discussion](https://groups.google.com/forum/#!topic/common-workflow-language/i76O6Edc_VU).

A bit of context/history:

One of the early design choices for CWL was to use JSON-LD because it would allow us, through RDF, to make use of all the nice related things that other people have already built and integrate with them. Examples include ontologies like EDAM, dcterms, wfdesc, DOAP, PROV-O and such.

One of the best things about JSON-LD is that it allows people who don't know anything about RDF to not care about it at all, and just treat it as regular JSON.

Almost immediately, we switched to using YAML since it's easier to read/write and supports comments. JSON is a subset of YAML (so was automatically supported), but JSON-LD is specific to JSON. That meant CWL files weren't exactly RDF since they first needed to be converted to JSON, but that seemed trivial.

We also wanted to have CWL workflows automatically be valid wfdesc workflows, but decided against it since the encoding was too verbose. We still re-used (extended) many wfdesc concepts and generating a `wfdesc:Workflow` from `cwl:Workflow` is easy.

To keep the CWL files DRY and maintainable, `import` and `include` directives were added. They are great and simplify things a lot, but it also meant that there was an additional step (loading references) before the CWL file can be converted to RDF.

We also chose to use Apache Avro, and one of the reasons was that its JSON representation can work with JSON-LD. However, that meant using the unintuitive syntax for optional values (e.g. `type: ["null", "string"]`). Draft-1 used JSON-Schema for defining types and restrictions on values, but wasn't possible to encode nicely in JSON-LD; the same issue (#171) exists for default values that are dicts/lists/files.

More recently, draft-3 specifies a simplified way of using identifiers (scoped, without leading `#`), which is amazing but also requires additional processing for RDF conversion.

The discussion that prompted this RFC was (among other things) about possibly extending the semantics of the JSON-LD `@context` property to reduce boilerplate, which would also need to be handled before conversion to RDF.

In short, we seem to be reducing the verbosity and boilerplate of CWL YAML files that was there due to  the initial idea of having CWL files be valid RDF and moving the complexity into a single pre-processing step that is currently handled by a separate project (Schema Salad) with own specification.
## Proposal

Personally, I love the fact that CWL is getting less verbose in each iteration. How about we go all the way? We can have a CWL-YAML syntax that has all the good stuff (able to convert to RDF and back, able to use other work through RDF) and almost none of the bad stuff (boilerplate, verbosity).

Let's see what we can change:
#### YAML vs RDF

We don't need to invent our RDF encoding. We can have the YAML syntax that works on its own and tools (and docs) for converting to RDF and back. The conversion would likely be done by restructuring the loaded YAML file, adding `@context` and interpreting as JSON-LD; but the conversion tools can work with other formats as well (Turtle, XML).

We just need to still allow people to use own RDF properties in the YAML encoding.
#### Remove @context

At the moment, this field is rarely even specified, and when specified - it's just boilerplate. However, when adding own properties with own prefix, one would add the prefixes to the `@context` array:

``` yaml
"@context":
 - http://wid.org/cwl/draft-3/context
 - {foaf: http://xmlns.com/foaf/0.1/}
```

Since base context only depends on CWL version, we can simplify this by replacing `@context` with an optional top-level field:

``` yaml
namespaces:
  foaf: http://xmlns.com/foaf/0.1/
```
#### Lists vs Dicts

A common pattern in draft-2 is to specify things as arrays with a required `id` field. This is because it's the JSON-LD version of the RDF "hasX" property pattern. If it only has one X (e.g. input), the value doesn't need to be a list. Lists are sometimes preferred for ordering, but in RDF semantics these are not ordered (not really lists).

In all of these cases, a dict/map is a preferred encoding (keys would be values of `id` property and values would omit the `id`) if the keys are short and have no special characters (which is the case in draft-3 scoped IDs).

So, instead of writing:

``` yaml
inputs:
  - id: reads
    type: File
  - id: reference
    type: File
```

We can write:

``` yaml
inputs:
  reads:
    type: File
  reference:
    type: File
```

This applies for tool/workflow/step inputs/outputs, steps, namespaces mentioned above and other things we might add later.
#### Optional inputs in Avro

`type: ["null", "string"]` is very unintuitive. Especially the quoted `"null"`. Wouldn't it be easier to just have a `required` field for the input (default of `false`)?

``` yaml
inputs:
  input1:
    type: File
    required: true
  input2:
    type: string
```

This would just apply for top-level input definitions. Any nested structures would still be standard Avro.
#### Workflow steps and wiring

Instead of repeating relevant inputs/outputs of the step implementation, we can simply omit outputs and introduce `connect` and `defaults` fields:

``` yaml
class: Workflow
inputs:
  file_inp: {type: File, required: true}
outputs:
  file_out:
    type: File
    connect: step2.outfile
steps:
  step1:
    run: {$import: myapp.cwl}
    connect: {input1: file_inp}
    defaults: {input2: 42}
  step2:
    run: {$import: otherapp.cwl}
    connect: {input1: step1.output1}  # No need to redeclare output1 in step1.
    defaults: {input2: [1, 2, 3]}  # Can have an array as default value.
```
#### Other things

These are the ones that have been bugging me the most. Do you have any of your own? What are unneeded complexities of CWL YAML encoding that can be represented in a better, isomorphic way?

**Comment by mr-c** [comment]
An important use case to consider for the current and future serializations is the need for automated generation of SDKs for reading and processing CWL files for Python, Java, et cetera..

**Comment by scottfrazer** [comment]
Very well written.

It's clear from your history/context section that syntax and readability is important.  If it weren't then CWL would probably probably have stayed as JSON and had an even more heavy-handed format that is meant to be handled by people.  Though since it is important that people can also easily read and write CWL, certain changes had to be made to accommodate that.  

> Personally, I love the fact that CWL is getting less verbose in each iteration. How about we go all the way? 

Yes!  I like all of the changes you suggested.  I think they're steps in the right direction.

I have to ask though, why not *really* go all the way.  Why not start with a blank text editor and drop preconceptions about what is feasible and what isn't and just create the syntax from scratch that you'd like to read and write?

Sure, I get that it's a disrupting change, but can this be at least considered for *some* future version of CWL?  I'm curious about what people perceive as the long-term strategy for CWL.  In 5 years with more resources and a few major versions from now, what does the syntax of CWL look like?  That's a discussion worth having too, so we can aim our sights in that direction.

I feel like the underlying assumption that the text format must be YAML or JSON really brings all syntax discussions to a grinding halt pretty quickly.  You only have so much wiggle room if you're constrained by YAML and it really doesn't let you wiggle enough toward feeling like a proper language.  

Perhaps another suggestion if an blank-slate syntax approach is off the table is mini trivially-parseable DSLs within YAML.  Using one of Nebojsa's examples:

```yaml
inputs:
  reads:
    type: File
  reference:
    type: File
```

Could be

```
inputs:
  - File reads
  - File reference
```

Honestly I'm not sure I'm even a big fan of this approach, but I thought it'd be worth mentioning.  I feel like if you're going to go that far, why not just ditch YAML altogether?

Another CWL syntax issue I have is with specifying command lines.  It's *really* hard to follow the `baseCommand` and `inputBinding` and `arguments` sections to construct a command line in my head.  Even with the proposed changes to template things a bit more it still feels a bit awkward.  I wish I had a suggestion instead of just a complaint... but maybe somebody else has an idea?

Also, @mr-c I totally agree with the need for automated generation of SDKs

**Comment by ntijanic** [comment]
> I have to ask though, why not really go all the way. Why not start with a blank text editor and drop preconceptions about what is feasible and what isn't and just create the syntax from scratch that you'd like to read and write?

I think https://gist.github.com/satra/9d8f8b8bf1ccbf89d5c8#file-grep-json was one such attempt, though likely still very primed.

There's few really nice things about yaml/json:
1. It's a well known base syntax.
2. No need for an envelope when transferring these files. A server can just serve a CWL json file and client can load it using regular json lib to look inside tool/wf metadata (label, versions, inputs, outputs etc).
3. It's trivial for people to extend it, either with compute-relevant things or with just custom metadata which can be included on nested objects such as inputs/outputs.

Though, a good DSL may enable 3) and make 1) not as important.

Adding further sugar to simplify some use cases would be great! The inputs example can possibly be further improved by something like:

```yaml
inputs:
  reference: File
  reads: File[]
  threshold: int?
  out_prefix:
    type: string?
```

i.e. if value of input is a string, interpret it as value for its `type` property, and add a shorthand for declaring types.

> Another CWL syntax issue I have is with specifying command lines. It's really hard to follow the baseCommand and inputBinding and arguments sections to construct a command line in my head. Even with the proposed changes to template things a bit more it still feels a bit awkward. I wish I had a suggestion instead of just a complaint... but maybe somebody else has an idea?

I like WDL's approach to this (template-like chunks in cmd line that include info from CWL's inputBinding). As mentioned in #176, we can use something like:

```
grep {ignore_case pref=-i} {max_count pref=-m sep='='} {pattern} {file}
```

> An important use case to consider for the current and future serializations is the need for automated generation of SDKs for reading and processing CWL files for Python, Java, et cetera..

How much can an SDK be generated? Perhaps I misunderstand, but it seems best we can do is something that Avro offers (describe CWL in Avro and such and let people generate DTOs).

**Comment by scottfrazer** [comment]
> 1. It's a well known base syntax.

I guess I feel that being 'well known' doesn't add a whole lot of value.  After all, XML is *very* well known but I'm sure nobody would agree that we should use XML

> 2. No need for an envelope when transferring these files. A server can just serve a CWL json file and client can load it using regular json lib to look inside tool/wf metadata (label, versions, inputs, outputs etc).

If I understand correctly, you're saying that you can serve CWL directly to clients because it's YAML?  Cromwell has a web service now that [accepts WDL](https://github.com/broadinstitute/cromwell#post-apiworkflowsversion) and calls `WdlNamespace.load()` on the WDL that people.  And our internal methods repository serves WDL directly... is that different than what you're saying?

--------------------------------------------------------

The issue I see with mixing in DSL-like features is that it undermines the claim that CWL is easy to parse because it's YAML.  Now, you have to say "It's YAML, but there are strings within the YAML that must be parsed in a custom way".  Avro then becomes just one step in the parsing process.  In the example you gave:

```
inputs:
  reference: File
  reads: File[]
  threshold: int?
  out_prefix:
    type: string?
```

As far as I know (correct me if I'm wrong), Avro won't be able to verify that `File[]` or `int?` is valid, so there will need to be custom parsing code to check this and turn it into the appropriate data structure.

With the command template as you've mentioned, there will likely need to be extra custom code that verifies that all curly-braces are matched, and the command template is well-formed.

My guess is that most people will probably choose not to try to write a parser for CWL themselves.  There's too much nit-picky parsing they have to do by hand (same is true of WDL!).  People will probably want to just do something like this:

```
import cwl
my_tool = cwl.parse("/path/to/file.cwl")
print(my_tool.command)
print(my_tool.command.instantiate(...))
```

And if that's the interface for parsing CWL, then who cares what is happening underneath?  Just like how every language has pretty much one accepted JSON parser that everybody uses and relies on, why can't the same be true of CWL?  And if that's the case, then why does the syntax need to suffer just so the implementers of the Python CWL module can use YAML and Avro?

> How much can an SDK be generated? Perhaps I misunderstand, but it seems best we can do is something that Avro offers.

We can definitely do better than Avro, in my opinion.  Avro is basically a parser that forces you into a JSON compatible syntax and it can't validate all of the DSL-like features that you are proposing.

However, there is a technology that CAN parse at a deeper level and does not force you into a JSON-compatible syntax.  The [grammar file for WDL](https://github.com/broadinstitute/wdl/blob/master/grammar.hgr) is 344 lines long and generates both a Java *and* Python Parser (both heavily unit tested).  This parser does much more than Avro does and it's also less code than the Avro schema.  It'll change parsing modes when it is parsing a command template, it'll handle escape sequences, and parse expressions using proper order of operations.

Granted, even a proper parser doesn't get you 100% of the way to a fully automatically generated SDK, but it does get you very far down that path.  The approach we take is that the parser itself is fully generated and that gets us about 85% of the way there, and then we hand-write the layer on top of the abstract syntax tree, which tends to be a lot shorter and simpler.

**Comment by tetron** [comment]
@ntijanic agree with all your suggestions except possibly the changes to wiring.  Specifically omitting 'outputs' breaks the direct relationship between the structure of the workflow document and the actual data flow it represents.

**Comment by tetron** [comment]
@scottfrazer CWL is defined primarily as an abstract document schema on the json data model.  This is a little different from thinking of it as a file format.  The best analogy I can give are homoiconic languages such as lisp and prolog where the code and data share the same structure, which enables much easier programmatic manipulation of the code.

A non-yaml, user friendly syntax that compiles to CWL is a very reasonable thing to build on top, because what matters for compatiblity is the deep structure and semantics, not the surface syntax.

**Comment by scottfrazer** [comment]
@tetron I knew we'd agree on something!  Homoiconicity is very important!  Unsurprisingly, WDL is homoiconic too!  "If a language is homoiconic, it means that the language text has the same structure as its abstract syntax tree (i.e. the AST and the syntax are isomorphic)"  That's *exactly* how WDL works.  You could construct the AST directly and then convert it to source code, and indeed that's how the WDL syntax highlighter and code formatter works:  each take an AST and rebuild source code.  You could take an AST, add some nodes onto it, then generate the corresponding source code.  If that's not homoiconicity, I don't know what is!

> CWL is defined primarily as an abstract document schema on the json data model. This is a little different from thinking of it as a file format.

I don't think it is different actually.  At the end of the day you have a string that needs to be a certain format.  If it does adhere to that format, you're returning the corresponding data structures.  What we're essentially talking about is "source code to data structures and vice versa" translation works.

> A non-yaml, user friendly syntax that compiles to CWL is a very reasonable thing to build on top

I have to disagree with this.  Once you go the route of making your own syntax, then you've obviated the need for CWL.  What is the very first data structure you get from a parser?  An AST of course... and why would you then bother to compile it to CWL just so you can re-parse it?  Why not just use the AST you already have?  It makes SO much more sense just to use the AST and not maintain this translation layer.  It feels to me like you're trying to sell this architecture to those who want a better syntax:

```
DSL source -> AST -> CWL -> JSON data model -> language API -> client code / engine
```

As opposed to the much simpler:

```
DSL source -> AST -> language API -> client code / engine
```

In the latter model it's much easier to make the first three layers all isomorphic.  The reason why I push so hard about this is that it's really not as simple as you think to just "compile to CWL".  I urge you to try it.  If you're going to have any flexibility or creativity with syntax, you will find it's very "square peg, round hole" to insert CWL in the process.  Trust me when I say that we when WDL started we really DID want to compile to CWL.  It'd be a win for everybody!  However, once code starts being written and you realize how far out of the way it is to first compile to CWL, it becomes obvious that CWL has to be adding a LOT of value to make it worth the effort.

**Comment by pgrosu** [comment]
@scottfrazer, maybe thinking of CWL as a universal communicator/translator might help.  It's like the ultimate connector between all languages/SDKs - we need standards to communicate (i.e. XML, HTML, JSON, etc.)  CWL is the most concise way of modeling the code and information, similar to how an abstract semantic graph (ASG) is a more concise and optimal representation of an abstract syntax tree (AST).

Hope it helps,
~p
